src/main/java/com/facebook/openwifirrm/ucentral/informationelement/Country.java

```diff
@@ -0,0 +1,173 @@
+ * Refer to the 802.11 specification for more details. Language in
+ * javadocs is taken from the specification.
+ */
+public class Country {
+
+	/** Defined in 802.11 */
+	public static final int TYPE = 7;
+
+	public static class CountryInfo {
+		/**
+		 * The lowest channel number in the CountryInfo.
+		 */
+		@SerializedName(
+			value = "First Channel Number"
+		) public final int firstChannelNumber;
+		/**
+		 * The maximum power, in dBm, allowed to be transmitted.
+		 */
+		@SerializedName(
+			value = "Maximum Transmit Power Level (in dBm)"
+		) public final int maximumTransmitPowerLevel;
+		/**
+		 * Number of channels this CountryInfo applies to. E.g., if First
+		 * Channel Number is 2 and Number of Channels is 4, this CountryInfo
+		 * describes channels 2, 3, 4, and 5.
+		 */
+		@SerializedName(
+			value = "Number of Channels"
+		) public final int numberOfChannels;
+
+		/**
+		 * Constructor. This constructor is currently never used, but it is
+		 * required by Java since at least one member variable is marked final.
+		 */
+		public CountryInfo(
+			int firstChannelNumber,
+			int maximumTransmitPowerLevel,
+			int numberOfChannels
+		) {
+			this.firstChannelNumber = firstChannelNumber;
+			this.maximumTransmitPowerLevel = maximumTransmitPowerLevel;
+			this.numberOfChannels = numberOfChannels;
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(
+				firstChannelNumber,
+				maximumTransmitPowerLevel,
+				numberOfChannels
+			);
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj) {
+				return true;
+			}
+			if (obj == null) {
+				return false;
+			}
+			if (getClass() != obj.getClass()) {
+				return false;
+			}
+			CountryInfo other = (CountryInfo) obj;
+			return firstChannelNumber == other.firstChannelNumber &&
+				maximumTransmitPowerLevel == other.maximumTransmitPowerLevel &&
+				numberOfChannels == other.numberOfChannels;
+		}
+
+		@Override
+		public String toString() {
+			return "CountryInfo [firstChannelNumber=" + firstChannelNumber +
+				", maximumTransmitPowerLevel=" + maximumTransmitPowerLevel +
+				", numberOfChannels=" + numberOfChannels + "]";
+		}
+	}
+
+	/**
+	 * Each constraint is a CountryInfo describing tx power constraints on
+	 * one or more channels, for the current country.
+	 */
+	@JsonAdapter(
+		ConstraintsDeserializer.class
+	) public final List<CountryInfo> constraints;
+
+	/**
+	 * Constructor. This constructor is currently never used, but it is
+	 * required by Java since at least one member variable is marked final.
+	 */
+	public Country(List<CountryInfo> countryInfos) {
+		this.constraints = Collections.unmodifiableList(countryInfos);
+	}
+
+	/**
+	 * ConstraintsDeserializer is used because the constraints field in the
+	 * json is "nested" (instead of a list of CountryInfos, it is a list of
+	 * JsonObjects each with exactly one key called "Country Info").
+	 */
+	private static class ConstraintsDeserializer
```

Comment:
...I am not sure how I feel about using gson to construct these IE classes, and seeing these custom deserializer classes especially makes me lean towards "no". The alternative is only *slightly* more code (ex. static method like `Country.parse(JsonObject contents) -> Country/null`) since manual parsing is so straightforward, and that might even end up being less code than a single custom deserializer class (and far easier to read/understand).

This is assuming we never need to re-serialize which is probably true.

Reply:
Ok, I will try a `static parse(JsonObject contents)` for every IE (except `HTOperationElement` and `VHTOperationElement` which are parsed and stored differently).